api/v1alpha3/virtualmachine_types.go

```diff
@@ -447,7 +447,7 @@ type VirtualMachineSpec struct {
 	MinHardwareVersion int32 `json:"minHardwareVersion,omitempty"`
 
 	// +optional
-	// +kubebuilder:validation:Format:=uuid4
+	// +kubebuilder:validation:Format:=uuid
```

Comment:
Wondering if we should just remove this validation and add our own. Unclear when this validation is actually applied, and does look correct for uuidv4. See: https://github.com/kubernetes/kubernetes/blob/master/vendor/k8s.io/kube-openapi/pkg/validation/strfmt/default.go#L64
See also: https://github.com/kubernetes/kubernetes/issues/97733

Reply:
IIRC this validation is applied by the K8s API Server when a CR is created or updated.
I spot-checked a few VM instance IDs and noticed that they are not actually in the format of UUID v4?

Reply:
> I spot-checked a few VM instance IDs and noticed that they are not actually in the format of UUID v4?

Can you expand on that?

I haven't seen the uuid4 validation fail when creating, only when updating with this MR.

The uuid package String() format used in vm-operator is always 8-4-4-4-12: 
```
xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx
```
See: https://github.com/google/uuid/blob/6e10cd1027e225e3ad7bfcc13c896abd165b02ef/uuid.go#L242

The regex for `uuid4` linked to above does not match, as it is 8-4-3-3-12

But the `uuid` format matches 8-4-4-4-12

And, kubernetes itself uses the same uuid package (8-4-4-4-12 format) for MetaData.UID:
https://github.com/kubernetes/kubernetes/blob/ac30b6a0e1572145964f320ceb9a18ed2c298e68/staging/src/k8s.io/apimachinery/pkg/util/uuid/uuid.go#L26

Reply:
I believe the K8s regex used for `uuid4` is, in fact, correct. 

> UUID4Pattern = `(?i)^[0-9a-f]{8}-?[0-9a-f]{4}-?4[0-9a-f]{3}-?[89ab][0-9a-f]{3}-?[0-9a-f]{12}$`

This is in the format of 8-4-4-4-12, with the third section starting with `4` and the fourth starting with `8`, `9`, `a`, or `b`.

The validation failed during upgrade probably because the `spec.biosUUID` field was being set from the existing `status.biosUUID` field in the `upgradeSchema()` function. However, the default instance UUID used by vSphere VM isn't UUID v4 (e.g. `503e71cd-ff35-cd68-b222-29b587ac954d`). So this update from controller was rejected by the API Server due to the `+kubebuilder:validation:Format:=uuid4` in our CRD.

To ensure backwards compatibility, we cannot have uuid4 validation for existing VMs.

Reply:
Ah, thank you, my eyes failed to parse the regex. Are you ok keeping the change to `uuid` validation?

> UUIDPattern = `(?i)^[0-9a-f]{8}-?[0-9a-f]{4}-?[0-9a-f]{4}-?[0-9a-f]{4}-?[0-9a-f]{12}$`

Reply:
Np! I am fine with keeping it as `uuid` to ensure this doesn't break existing VMs with default UUIDs that are not v4. If we want to enforce UUID v4 for new VMs, we could leverage validating webhook to check this field during new VM creation.